common/functions/src/scalars/conditionals/if.rs

```diff
@@ -41,102 +42,330 @@ impl IfFunction {
+                    return Ok(builder.build(input_rows));
+                }else {
+                    for (predicate, (row, r_val)) in iter {
+                        if predicate {
+                            builder.append(l_val, true);
+                        }else {
+                            builder.append(r_val, rhs_viewer.valid_at(row));
+                        }
+                    }
+                    return Ok(builder.build(input_rows));
+                }
+            }, {
+                unimplemented!()
+            });
+        } else if rhs.is_const() {
+            // rhs is const
+            with_match_scalar_type!(type_id.to_physical_type(), |$T| {
+                let left_viewer = $T::try_create_viewer(&lhs)?;
+                let l_val = left_viewer.value_at(0);
+                let right_viewer = $T::try_create_viewer(&rhs)?;
+                let r_val = right_viewer.value_at(0);
+
+                if reverse {
+                    let iter = cond_col.iter().map(|predicate| if predicate { r_val } else { l_val });
+                    return Ok(Arc::new(ColumnBuilder::<$T>::from_iterator(iter)));
+                }else {
+                    let iter = cond_col.iter().map(|predicate| if predicate { l_val } else { r_val });
+                    return Ok(Arc::new(ColumnBuilder::<$T>::from_iterator(iter)));
+                }
+            }, {
+                unimplemented!()
+            });
         } else {
-            let boolean_dt = BooleanType::arc();
-            predicate = cast_column_field(&columns[0], &boolean_dt)?;
+            // rhs is primitive type
+            macro_rules! build_with_primitive_type {
+                ($column:ident, $T:ident) => {
+                    let left_viewer = $T::try_create_viewer(&lhs)?;
+                    let l_val = left_viewer.value_at(0);
+                    let rhs = $T::try_create_viewer(&rhs)?;
+
+                    if reverse {
+                        let iter = cond_col.iter().zip(rhs.iter()).map(|(predicate, r_val)| {
+                            if predicate {
+                                r_val
+                            } else {
+                                l_val
+                            }
+                        });
+                        let col = <$T as Scalar>::ColumnType::from_iterator(iter);
+                        return Ok(col.arc());
+                    } else {
+                        let iter = cond_col.iter().zip(rhs.iter()).map(|(predicate, r_val)| {
+                            if predicate {
+                                l_val
+                            } else {
+                                r_val
+                            }
+                        });
+                        let col = <$T as Scalar>::ColumnType::from_iterator(iter);
+                        return Ok(col.arc());
+                    }
+                };
+            }
+
+            match type_id {
+                TypeID::Boolean => {
```

Comment:
There is binding from `Scalar::ColumnType --->   Column`

Reply:
resolved